sdk/core/core/src/az_json_builder.c

```diff
@@ -0,0 +1,116 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#include <az_json_builder.h>
+
+#include <az_str.h>
+
+#include <_az_cfg.h>
+
+AZ_NODISCARD az_result
+az_json_builder_init(az_json_builder * const out, az_span_action const write) {
+  AZ_CONTRACT_ARG_NOT_NULL(out);
+
+  *out = (az_json_builder){ .write = write };
+
+  return AZ_OK;
+}
+
+AZ_NODISCARD static az_result
+az_json_builder_write_str(az_json_builder * const self, az_span const value) {
+  AZ_CONTRACT_ARG_NOT_NULL(self);
+
+  az_span_action const write = self->write;
+  AZ_RETURN_IF_FAILED(az_span_action_do(write, AZ_STR("\"")));
+  AZ_RETURN_IF_FAILED(az_span_action_do(write, value));
+  AZ_RETURN_IF_FAILED(az_span_action_do(write, AZ_STR("\"")));
+  return AZ_OK;
+}
+
+AZ_NODISCARD az_result
+az_json_builder_write(az_json_builder * const self, az_json_value const value) {
+  AZ_CONTRACT_ARG_NOT_NULL(self);
+
+  az_span_action const write = self->write;
+  switch (value.kind) {
+    case AZ_JSON_VALUE_NULL: {
+      self->need_comma = true;
+      return az_span_action_do(write, AZ_STR("null"));
+    }
+    case AZ_JSON_VALUE_BOOLEAN: {
+      self->need_comma = true;
+      return az_span_action_do(write, value.data.boolean ? AZ_STR("true") : AZ_STR("false"));
+    }
+    case AZ_JSON_VALUE_STRING: {
+      self->need_comma = true;
+      return az_json_builder_write_str(self, value.data.string);
+    }
+    case AZ_JSON_VALUE_NUMBER: {
+      self->need_comma = true;
+      // TODO:
+      return AZ_ERROR_NOT_IMPLEMENTED;
+    }
+    case AZ_JSON_VALUE_OBJECT: {
+      self->need_comma = false;
```

Comment:
If I understand the 'need_comma' thing right, I think you always want a comma unless the last character in the span is a { or a [, right? If so, kill the 'need_comma' field and just look at the last character in the span; this will also kill all the code that sets the field.

Reply:
Currently, I'm using polymorphic writer which can't return back the last character. I may use it if I use `az_span_builder` instead of `write` but even in this case we would like to keep `az_span_builder` as a black box with limited set of functions.